cranelift/codegen/src/ir/memtype.rs

```diff
@@ -0,0 +1,171 @@
+//! can use in proof-carrying code to validate accesses to structs and
+//! propagate facts onto the loaded values as well.
+//!
+//! Memory types are meant to be rich enough to describe the *layout*
+//! of values in memory, but do not necessarily need to embody
+//! higher-level features such as subtyping directly. Rather, they
+//! should encode an implementation of a type or object system.
+//!
+//! Note also that it is a non-goal for now for this type system to be
+//! "complete" or fully orthogonal: we have some restrictions now
+//! (e.g., struct fields are only primitives) because this is all we
+//! need for existing PCC applications, and it keeps the
+//! implementation simpler.
+//!
+//! There are a few basic kinds of types:
+//!
+//! - A struct is an aggregate of fields and an overall size. Each
+//!   field has a *primitive Cranelift type*. This is for simplicity's
+//!   sake: we do not allow nested memory types because to do so
+//!   invites cycles, requires recursive computation of sizes, creates
+//!   complicated questions when field types are dynamically-sized,
+//!   and in general is more complexity than we need.
+//!
+//!   The expectation (validated by PCC) is that when a checked load
+//!   or store accesses memory typed by a memory type, accesses will
+//!   only be to fields at offsets named in the type, and will be via
+//!   the given Cranelift type -- i.e., no type-punning occurs in
+//!   memory.
+//!
+//!   The overall size of the struct may be larger than that implied
+//!   by the fields because (i) we may not want or need to name all
+//!   the actually-existing fields in the memory type, and (ii) there
+//!   may be alignment padding that we also don't want or need to
+//!   represent explicitly.
+//!
+//! - A static memory is an untyped blob of storage with a static
+//!   size. This is memory that can be accessed with any type of load
+//!   or store at any valid offset.
+//!
+//!   Note that this is *distinct* from an "array of u8" kind of
+//!   representation of memory, if/when we can represent such a thing,
+//!   because the expectation with memory types' fields (including
+//!   array elements) is that they are strongly typed, only accessed
+//!   via that type, and not type-punned. We don't want to imply any
+//!   restriction on load/store size, or any actual structure, with
+//!   untyped memory; it's just a blob.
+//!
+//! Eventually we plan to also have:
+//!
+//! - A dynamic memory is an untyped blob of storage with a size given
+//!   by a global value (GV). This is otherwise just like the "static
+//!   memory" variant described above.
+//!
+//! - A dynamic array is a sequence of struct memory types, with a
+//!   length given by a global value (GV). This is useful to model,
+//!   e.g., tables.
+//!
+//! - A discriminated union is a union of several memory types
+//!   together with a tag field. This will be useful to model and
+//!   verify subtyping/downcasting for Wasm GC, among other uses.
```

Comment:
Note for the future that wasm gc will also require nullability, which almost but doesn't quite fit into the idea of discriminant tags.

Reply:
Added another note for this in the doc-comment!